### PR TITLE
compilers: Every compiler can run code

### PR DESCRIPTION
**Pull request by xclaesse**
Those are commits extracted from one of my Rust WIP branches because they can be merged separately.

### CODE REVIEW

File: mesonbuild/compilers/compilers.py [hunk 1]
```diff
@@ -871,12 +871,7 @@ def compile(self, code: 'mesonlib.FileOrString',
                 no_ccache = True
                 contents = code
             else:
-                srcname = code.fname
-                if not is_object(code.fname):
-                    with open(code.fname, encoding='utf-8') as f:
-                        contents = f.read()
-                else:
-                    contents = '<binary>'
+                contents = srcname = code.fname
```
Comment (eli-schwartz) [inline mesonbuild/compilers/compilers.py:875]: I'd prefer to also change the log message here to be more clear about what it's doing.

File: mesonbuild/compilers/compilers.py [hunk 2]
```diff
@@ -717,7 +717,8 @@ def has_header_symbol(self, hname: str, symbol: str, prefix: str,
 
     def run(self, code: 'mesonlib.FileOrString', env: 'Environment', *,
             extra_args: T.Union[T.List[str], T.Callable[[CompileCheckMode], T.List[str]], None] = None,
-            dependencies: T.Optional[T.List['Dependency']] = None) -> RunResult:
+            dependencies: T.Optional[T.List['Dependency']] = None,
+            run_cwd: T.Optional[str] = None) -> RunResult:
```
Comment (eli-schwartz) [inline mesonbuild/compilers/compilers.py:721]: Commit message: what is "1sftworkdir"? :D

File: mesonbuild/compilers/compilers.py (hunk 1 above)
Comment (xclaesse) [reply mesonbuild/compilers/compilers.py]: done

File: mesonbuild/compilers/compilers.py (hunk 2 above)
Comment (xclaesse) [reply mesonbuild/compilers/compilers.py]: fixed

File: mesonbuild/interpreter/compiler.py [hunk 3]
```diff
@@ -295,6 +295,11 @@ def alignment_method(self, args: T.Tuple[str], kwargs: 'AlignmentKw') -> int:
     @typed_pos_args('compiler.run', (str, mesonlib.File))
     @typed_kwargs('compiler.run', *_COMPILES_KWS)
     def run_method(self, args: T.Tuple['mesonlib.FileOrString'], kwargs: 'CompileKW') -> 'RunResult':
+        from ..compilers.d import DCompiler
+        from ..compilers.mixins.clike import CLikeCompiler
+        if not isinstance(self.compiler, (CLikeCompiler, DCompiler)):
```
Comment (dcbaker) [inline mesonbuild/interpreter/compiler.py:300]: I thought the general move was away from this kind of import and isinstance checking. could we just do:
```python
if self.compiler.language not in {'d', 'c', 'cpp', 'objc', 'objcpp'}:
     FeatureNew.single_use(...)
```

Comment (xclaesse) [reply mesonbuild/interpreter/compiler.py]: Good point, changed.

**Review by dcbaker — APPROVED** [review]

**Comment by xclaesse** [comment]
@dcbaker I had to add an extra commit to make mypy happy, and that's actually a big cleanup.

**Review by dcbaker — APPROVED** [review]

**Comment by xclaesse** [comment]
> Test failures seem related.

Very surprising, but yes. There are actually 2 issues:
- Rust compiler sanity check skips native-static-libs part if it cannot run host binaries
- Even in a native build, environment.need_exe_wrapper() returns True during compiler sanity checks because that happens before _redetect_machines().

**Comment by eli-schwartz** [comment]
From a quick glance this seems to be the same as #12481, should that one be closed?

**Comment by xclaesse** [comment]
> From a quick glance this seems to be the same as #12481, should that one be closed?

Opened it twice indeed. Closed the other one.

**Comment by xclaesse** [comment]
CI passed now

**Pull request merged**